### ISSUE TITLE
J-lens read path: lens resolution, jlens_read, jlens_concept_probe, jlens_sparse_inventory

### ISSUE THREAD

**Issue by speediedan** [speediedan/interpretune#419]
## Summary

Sub-issue of #225. Introduce the J-lens **read** surface in-tree: the artifact resolution the ops need, plus the three readout ops the workspace paper's analyses are built on. Today interpretune has the J-space **write** path (`patch`, #320) and nothing that can produce a readout, so every representation-level probe designed for this technique is blocked.

## Why this is the largest gap against the paper

[Gurnee et al. 2026](https://transformer-circuits.pub/2026/workspace/) is about what the workspace *holds*: `lens(h_l) = softmax(W_U . norm(J_l h_l))` gives the ranked vocabulary tokens an activation is disposed to make the model say. Almost everything else in the paper (which layers carry a concept, whether an unverbalized intermediate is present before it is swappable, how much of a vector lives in the J-space) is a readout or a decomposition over that readout. Interpretune can currently steer in a basis it cannot read.

Blocked on this issue: per-feature J-space signatures, the workspace-flip-vs-logit-flip ordering readout and the two-hop battery, the polysemy-cancellation probe designed in `src/it_examples/experiments/notebook/concept_direction/concept_direction_analysis.md`, and the vector construction a `basis="jlens"` selector should share.

## Scope

- [ ] **Lens artifact resolution in the sanctioned namespace.** Resolve pre-fitted lenses from [`neuronpedia/jacobian-lens`](https://huggingface.co/neuronpedia/jacobian-lens) (39 models, MIT): lazy per-layer `J_l`, cached under an `IT_NP_CACHE`-style location, plus a layer-percentile to fitted-layer helper. **Place it in `interpretune.analysis.optools` beside `resolve_unembed_and_norm_scale`**: hub op collections may bind importable parameters only to their own repo modules or that namespace, so putting it there is what lets `speediedan/jlens_steering_ops` migrate onto it (#273) rather than keep a private copy.
- [ ] **Probe the repo layout rather than formatting a path.** `{model}/jlens/Salesforce-wikitext/{model}_jacobian_lens.pt` holds for most entries and not all: `gpt2-small/` holds `gpt2_jacobian_lens.pt`, `gemma-3-1b/` holds `gemma-3-1b-pt_jacobian_lens.pt`, `gemma-4-e2b/` holds `gemma-4-E2B_...`, and `qwen3.5-4b/` carries an `_n1000` variant beside the default. The collection works around this with `jlens_model_id` / `jlens_lens_path` overrides; the in-tree loader should discover, and name what it could not resolve.
- [ ] **`jlens_read`**: ranked tokens at a set of (layer, position) pairs, readout-faithful (the final norm applied as the readout applies it, per #330). Expose whether the input-dependent `1/rms` scalar is included: it changes no ranking *within* a position and does change magnitude comparisons *across* positions, so this is a documented flag rather than a silent choice.
- [ ] **`jlens_concept_probe`**: cosine of an activation against chosen concept vectors, using the folded construction `v_c = (W_U[c] * s) @ J_l`. #330 section 2 derives that this is the direction the readout computes; the paper's "rows of `W_U J`" shorthand drops the norm and is exact only when `s = 1`.
- [ ] **`jlens_sparse_inventory`**: the sparse nonnegative decomposition over the J-lens dictionary by gradient pursuit, `k` configurable (the paper observes roughly 25 simultaneously active). This is the op that makes "the J-space component of this vector" computable at all, and it is the paper's central quantitative move (J-space component versus remainder); nothing upstream surfaces it.
- [ ] All three behind the `AnalysisBackend` seam with TransformerLens and NNsight parity, as the circuit-tracer ops are. No backend-specific imports in the bundled op definitions.
- [ ] Tests: synthetic-lens CPU fixtures for the ops' math and schemas; a cross-backend parity case; one `RunIf(cuda)` smoke against a real gemma lens.

## The #225 / #273 boundary this touches

The collection must keep something real to publish (#225's stated boundary). Proposal, to confirm before landing: **in-tree owns the lens surface** (resolution, readout, probe, decomposition); **the collection owns steering recipes** (the concept pole-pair construction with its measured folding option, and anything task-fitted). Landing the loader in `optools` is what makes that split cooperative rather than duplicative.

## Notes

- Artifacts are small: 66 MB for gemma-3-1b-it, 265 MB for gemma-2-2b, 432 MB for gemma-3-4b.
- The reference implementation [`anthropics/jacobian-lens`](https://github.com/anthropics/jacobian-lens) (Apache-2.0) covers HF models with `from_pretrained` / `apply` / `fit` / `merge` and is declared unmaintained and not accepting contributions. Treat it as a correctness reference, not a dependency to take: it has no TransformerLens path and no batch or op integration.
- The NNsight resolver's architecture registry covers GPT2, Gemma2, Gemma3 and Llama today; models outside it should fail with a named error rather than a wrong answer.

## Relationships

- #225 (parent, in-tree half) - #273 (the collection that migrates onto the loader)
- Blocks the basis selector, per-feature signatures, the experimental battery and the demo refresh

**Comment by speediedan** [comment]
## Design, against a measured layout rather than an assumed one (2026-09-06)

The boundary this issue proposed is **confirmed by the maintainer**: in-tree owns the lens surface
including the loader in `interpretune.analysis.optools`, the collection owns steering recipes. So the
loader lands here and the collection migrates onto it.

Before designing the resolver I inventoried the repository, and the scope item above understates the
problem in a way that changes the design.

### The canonical path is the minority case, not the exception

`{model}/jlens/Salesforce-wikitext/{model}_jacobian_lens.pt` matches **13 of 40 artifacts**. Twenty-seven
do not. The published collection formats exactly that string, so it resolves about a third of the
repository unaided, which is why it needs `jlens_model_id` / `jlens_lens_path` overrides as often as it
does.

The deviations are not one family of typo. Measured across 39 model directories:

| deviation | example | count |
| --- | --- | --- |
| stem is the HF name, not the directory | `gpt2-small/` holds `gpt2_...`; `llama3.1-8b-it/` holds `Llama-3.1-8B-Instruct_...` | most |
| case differs from the directory | `gemma-4-e2b/` holds `gemma-4-E2B_...` | 3 |
| base-model suffix differs | `gemma-3-1b/` holds `gemma-3-1b-pt_...`; `qwen3.5-2b-pt/` holds `Qwen3.5-2B-Base_...` | several |
| a fitting-run variant suffix | `..._jacobian_lens_n1000.pt`, `..._jacobian_lens_cblank.pt` | 3 |
| the fitting corpus is not wikitext | `deepseek-v4-flash/jlens/NeelNanda-pile-10k/` | 1 |

Two shapes break the obvious repairs specifically:

- **`qwen3.5-4b` carries two artifacts** (default and `_n1000`), so "take the only `.pt` under the
  directory" is wrong there.
- **`qwen3.6-27b` and `deepseek-v4-flash` carry only a suffixed artifact**, so "require the unsuffixed
  name" is wrong there.

Any rule that is not discovery fails one of those two.

### The sidecar makes this a lookup rather than a heuristic

Every model directory carries a `config.yaml` beside the artifact, and it holds the mapping the resolver
actually needs:

```yaml
np_model_id: gpt2-small
hf_model_name: openai-community/gpt2
dataset: {name: Salesforce/wikitext, config: wikitext-103-raw-v1, ...}
fit:     {n_prompts: 1000, dtype: bfloat16, stop_at_delta: 0.002, ...}
results: {prompts_fitted: 277, final_identity_distance: 1.305569, final_mean_rel_change: 0.00159806}
```

So the design is **the heuristic proposes and the sidecar disposes**. Normalize the loaded model's HF id
to a candidate directory, then read that directory's `config.yaml` and require `hf_model_name` to match
before loading anything. If it does not match, scan the sidecars (39 files, each a couple of KB) and
resolve properly. This matters more than a tidiness argument: a mismatched lens produces a readout that
is entirely plausible and quietly wrong, which is the failure mode this repository already treats as the
one worth engineering against. A name-formatted path cannot detect it; a verified one cannot miss it.

Selection when a directory holds several artifacts: prefer the unsuffixed stem, take a lone candidate
whatever its suffix, and **raise naming every candidate** when there are several and none is unsuffixed.
Never pick by sort order.

### Surface

```python
# interpretune.analysis.optools
JLensArtifact          # NamedTuple: j_by_layer, source_layers, d_model, n_prompts, provenance
resolve_jlens(module, *, repo_id=..., model_id=None, path=None, revision=None) -> JLensArtifact
jlens_layer_for_percentile(artifact, percentile) -> int
```

`resolve_jlens` takes the same explicit escapes the collection has today (`model_id`, `path`), so a user
with a lens the resolver cannot place is never blocked, and the resolver reports what it could not
resolve by name. Checkpoint format is `{"J": {layer: Tensor[d, d]}, "n_prompts": int,
"source_layers": [int], "d_model": int}`; per-layer tensors load lazily, since a 27B lens is large and a
readout usually wants one layer.

**Provenance is part of the return value, not a debug aid.** `results.prompts_fitted` is 277 for gpt2
against a nominal `n_prompts: 1000`, because fitting stops on a convergence delta. So "fit over about a
thousand prompts" is the configured ceiling and not what happened, and it varies per model.
`final_identity_distance` (1.31 for gpt2) says how far the lens is from the identity, which is the
degenerate case where a J-lens reduces to a logit lens. A probe reporting a weak result against a lens
that barely left the identity is reporting something different from a weak result against a
well-separated one, and the consumer can only tell if the number travels with the artifact.

### The three ops

- **`jlens_read`**: ranked tokens at (layer, position) pairs. `include_rms_scale` is an explicit flag
  because the `1/rms(x)` factor changes no ranking within a position and does change magnitude
  comparisons across positions. Defaulting it silently would make cross-position comparisons wrong
  exactly when someone starts making them.
- **`jlens_concept_probe`**: cosine against concept vectors built with the readout-faithful
  construction. That construction is now settled and centralized: `fold_norm_into_unembed_rows` applies
  the elementwise scale, and for LayerNorm models the centering projector the readout implies, with the
  derivation in `docs/jlens_norm_folding.md`. The probe composes `rows @ J` and does not re-derive it.
- **`jlens_sparse_inventory`**: sparse nonnegative decomposition over the lens dictionary by gradient
  pursuit, `k` configurable. This is the op that makes "the J-space component of this vector"
  computable, so it needs a reconstruction-residual output alongside the coefficients; a decomposition
  reported without its residual cannot distinguish a faithful account from a poor one.

All three sit behind the `AnalysisBackend` seam with TransformerLens and NNsight parity, with no
backend imports in the bundled definitions. Each declares `required_capabilities` and, where it
constrains them, `required_position_scopes`, so an incompatible target refuses by name before anything
runs rather than after earlier ops in a composition have executed.

### Tests

Synthetic-lens CPU fixtures for math and schemas; a cross-backend parity case; one CUDA smoke against a
real gemma lens. Plus two the layout measurement above argues for directly, both CPU and offline against
a recorded listing: the resolver places every one of the 39 directories, and it **raises rather than
guessing** on the two-artifact directory. The second is the one that would catch a regression to
sort-order selection, which is the repair someone will reach for.

### Not in scope here

The fitting corpus varies (one model is not wikitext), and nothing in the surface above lets a caller ask
for a lens by corpus. That is a real limitation with one instance today, so it is stated rather than
designed around; if a second corpus appears for a model that already has one, the selection rule above
starts guessing and needs revisiting.

**Comment by speediedan** [comment]
### PR #482 lands the read path. This issue stays open until the CUDA smoke is green (2026-09-06)

Delivered against this issue's scope: the lens resolver in `interpretune.analysis.optools`, all three
ops (`jlens_read`, `jlens_concept_probe`, `jlens_sparse_inventory`), and the test surface. Stacked on
the folding PR, which the concept probe consumes.

**The layout item was understated in the body above and is now measured.** The canonical path holds for
**13 of 40** published artifacts, not "most". Resolution therefore discovers and verifies against the
`config.yaml` sidecar rather than formatting a string. Full breakdown is in the design comment above.

**Remaining before this closes:**

- [ ] The CUDA smoke against the published gpt2 lens completes green. It is written and has run twice;
      both runs failed, each for a real defect it found (a short-loaded model resolving no lens, and a
      device mismatch between CPU activations and a model-resident unembed). Both are fixed in the PR;
      the re-run is queued behind the GPU lease. **Not claiming this one until it passes.**
- [ ] Cross-backend parity for the concept probe and the sparse inventory. The readout has a parity
      case; the other two do not yet.

**Obviated rather than delivered:** the scope item asking for a layer-percentile helper "plus a
`{model}/jlens/...` path formatter" is delivered only in its first half, deliberately. A formatter is
what the measurement above argues against, so it exists nowhere in this PR, and `model_id=` / `path=`
cover the case where a caller knows better than the resolver.

**One thing the sidecars settled that reaches past this issue.** They record what fitting actually did,
and it is not what the configured parameters say: gpt2's lens reports `prompts_fitted: 277` against
`n_prompts: 1000`, because fitting stops on a convergence delta. Any prose that describes these lenses
as fit over roughly a thousand prompts is quoting a ceiling. The count and the lens's
`final_identity_distance` now travel on the resolved artifact so a consumer can weigh a result against
the quality of the lens that produced it.

**Comment by speediedan** [comment]
### The CUDA smoke is green (2026-09-06)

Third run, passing: 18 collected, 1 selected, 1 passed, against the fixed code (the fixes were
authored 40 minutes before the run, and the run immediately before it failed on exactly the device
error one of them addresses).

So of the two items left on this issue when #482 opened:

- [x] The CUDA smoke against the published gpt2 lens completes green.
- [ ] Cross-backend parity for the concept probe and the sparse inventory. The readout has a parity
      case; these two do not, so this issue stays open after #482 merges.

Worth recording what the green does not reach, since a narrow one reads as broad: the smoke exercises
one lens, one model, one device. The two-artifact refusal, the sidecar-less directory, the
non-wikitext corpus and the LayerNorm centering path are covered only by the recorded fixture, which
is the same instrument that failed to detect both defects this smoke found.

**Comment by speediedan** [comment]
Scope verified fully landed on main (6cd1b9ea), per the INSTRUCT-20 check. Evidence, all read off origin/main: (1) resolution in the sanctioned namespace — resolve_jlens, resolve_jlens_layer, jlens_direction_rows, jlens_layer_for_percentile in analysis/optools.py, with repo-layout discovery; (2) all three ops live with impl + YAML + top-level wrappers + stubs — jlens_read, jlens_concept_probe, jlens_sparse_inventory in analysis/ops/bundled/jlens/ (landed #482 7ecee4b0, extended by #559 norm-scale-from-module, #544 basis unification, #568 concept_basis); (3) readout-faithful with the documented jlens_include_rms_scale flag; probe uses the folded construction; (4) backend-seam clean — the op file imports only torch/transformers/base/optools, all backend specifics flow through resolve_unembed_and_norm_scale(module) + resolve_jlens_layer; (5) tests — 18 read-op + 14 resolution cases incl. TL-vs-HF cross-backend agreement (standalone) and a real-lens CUDA smoke (min_cuda_gpus=1). No remainder; closing.
